// @sap/cds-runtime/lib/cds-services/services/handlers/onDraftEdit.js
const generateUUID = require('../../../common/utils/uuid')
const { getCompositionTree } = require('../utils/compositionTree')
const getColumns = require('../utils/columns')
const { getTransition } = require('../../../common/utils/resolveView')
const {
  draftIsLocked,
  ensureDraftsSuffix,
  ensureNoDraftsSuffix,
  getDraftName,
  getEntityName,
  getServiceNamespace,
  getSubCQNs,
  setStatusCodeAndHeader
} = require('../utils/draftUtils')
const { isActiveEntityRequested, getKeyData } = require('../utils/draftWhereUtils')

const _getDraftColumns = draftUUID => ({
  IsActiveEntity: false,
  HasDraftEntity: false,
  HasActiveEntity: true,
  DraftAdministrativeData_DraftUUID: draftUUID
})

const _getAdminData = ({ user }, draftUUID, time) => {
  const currentUser = user.id || null
  return {
    DraftUUID: draftUUID,
    CreationDateTime: time,
    CreatedByUser: currentUser,
    LastChangeDateTime: time,
    LastChangedByUser: currentUser,
    DraftIsCreatedByMe: true,
    DraftIsProcessedByMe: true,
    InProcessByUser: currentUser
  }
}

const _getInsertAdminDataCQN = ({ statements: { INSERT }, user }, draftUUID, time) => {
  return INSERT.into('DRAFT.DraftAdministrativeData').entries(_getAdminData({ user }, draftUUID, time))
}

const _getLockWhere = (where, columnsMap) => {
  if (columnsMap.size === 0) {
    return where
  }

  const whereKeys = Object.keys(where)
  const lockWhere = {}

  whereKeys.forEach(key => {
    const lockKey = columnsMap.get(key).ref[0]
    lockWhere[lockKey] = where[key]
  })

  return lockWhere
}

const _select = async (CQNs, context) => {
  let allResults

  try {
    allResults = await Promise.all(CQNs.map(CQN => context.run(CQN)))
  } catch (err) {
    // resource busy and NOWAIT (WAIT 0) specified (heuristic error handling method)
    if (err.query.includes('FOR UPDATE')) {
      context.reject(409, 'DRAFT_ALREADY_EXISTS')
    }

    context.reject(err)
  }

  return allResults
}

/**
 * Generic event handler for draft edit requests.
 */
const _handler = async function (req) {
  if (!isActiveEntityRequested(req.query.SELECT.where || [])) {
    req.reject(400)
  }

  const { definitions } = this.model

  // TODO replace with generic where filter
  const keys = Object.keys(req.target.keys).filter(key => key !== 'IsActiveEntity')
  const data = getKeyData(keys, req.query.SELECT.from.ref[0].where)
  const rootWhere = keys.reduce((res, key) => {
    res[key] = data[key]
    return res
  }, {})
  const transition = getTransition(req.target)
  const lockWhere = _getLockWhere(rootWhere, transition.mapping)

  // gets the underlying target entity, as record locking can't be
  // applied to localized views
  const lockTargetEntity = transition.target

  // Lock the root record of the active entity to prevent simultaneous access to it,
  // thus preventing duplicate draft entities from being created or overwritten.
  // Only allows one active entity to be processed at a time, locking out other
  // users who need to edit the same record simultaneously.
  // .forUpdate(): lock the record, a wait of 0 is equivalent to no wait
  const lockRecordCQN = req.statements.SELECT.from(lockTargetEntity)
    .where(lockWhere)
    .forUpdate({ wait: 0 })

  const columnNames = getColumns(req.target, true)
  const rootCQN = req.statements.SELECT.from(req.target, columnNames).where(rootWhere)
  const subCQNs = getSubCQNs({
    definitions,
    context: req,
    rootCQN,
    compositionTree: getCompositionTree({ definitions, rootEntityName: ensureNoDraftsSuffix(req.target.name) })
  })
  const rootDraftName = ensureDraftsSuffix(req.target.name)
  const draftExistsCQN = req.statements.SELECT.from(rootDraftName, [
    'DraftAdministrativeData_DraftUUID as DraftUUID'
  ]).where(rootWhere)
  const selectCQNs = [rootCQN, ...subCQNs.map(obj => obj.cqn)]
  const lockAndSelectCQNs = [lockRecordCQN, draftExistsCQN, ...selectCQNs]
  const [, draftExists, ...results] = await _select(lockAndSelectCQNs, req)

  if (!results[0].length) {
    req.reject(404)
  }

  if (draftExists.length) {
    const adminData = await req.run(
      req.statements.SELECT.one('DRAFT.DraftAdministrativeData', ['InProcessByUser', 'LastChangeDateTime']).where(
        draftExists[0]
      )
    )

    // draft is locked (default cancellation timeout timer has not expired) OR
    // draft is not locked but must be rejected for popup
    if (draftIsLocked(adminData.LastChangeDateTime) || req.data.PreserveChanges) {
      req.reject(409, 'DRAFT_ALREADY_EXISTS')
    }

    await Promise.all([
      req.run(req.statements.DELETE.from('DRAFT.DraftAdministrativeData').where(draftExists[0])),
      req.run(req.statements.DELETE.from(rootDraftName).where(rootWhere))
    ])
  }

  const time = new Date().toISOString()
  const draftUUID = generateUUID()
  const insertCQNs = [_getInsertAdminDataCQN(req, draftUUID, time)]
  const serviceNamespace = getServiceNamespace(ensureNoDraftsSuffix(req.target.name))
  const entityName = getEntityName(ensureNoDraftsSuffix(req.target.name))

  for (const resultIndex in results) {
    if (results[resultIndex].length === 0) continue
    const draftEntity = getDraftName(serviceNamespace, getEntityName(selectCQNs[resultIndex].SELECT.from.ref[0]))
    const entries = results[resultIndex].map(entityResult =>
      Object.assign({}, entityResult, _getDraftColumns(draftUUID))
    )

    insertCQNs.push(req.statements.INSERT.into(draftEntity).entries(entries))
  }

  await Promise.all(insertCQNs.map(CQN => req.run(CQN)))
  setStatusCodeAndHeader(req._.odataRes, rootWhere, entityName, false)

  return Object.assign({}, results[0][0], { HasDraftEntity: false, HasActiveEntity: true, IsActiveEntity: false })
}

const { ODATA, COMMON } = require('../../../common/constants/annotation')
const _relevant = e => e[ODATA.DRAFT] || e[COMMON.DRAFT_NODE.PREP_ACTION]

module.exports = function () {
  for (const entity of Object.values(this.entities).filter(_relevant)) {
    this.on('EDIT', entity, _handler)
  }
}
