// @sap/cds/lib/env/defaults.js
module.exports = {

  features: {
    fiori_preview: true,
    skip_unused: true,
    one_model: true,
    snapi: true,
    cls: true,
    localized: true,
    skip_reclassification: false,
    journal: false
  },
  'log-level': 'warn',

  folders: { // IMPORTANT: order is significant for cds.load('*')
    db: 'db/',
    srv: 'srv/',
    app: 'app/',
  },

  i18n: {
    folders: ['_i18n', 'i18n', 'assets/i18n'],
    for_sqlite: ['de', 'fr'],
    languages: 'all', // or array.  'all': whatever language files are found next to models
    default_language: 'en',
    preserved_locales: [
      'en_GB',
      'es_CO',
      'es_MX',
      'fr_CA',
      'pt_PT',
      'zh_CN',
      'zh_HK',
      'zh_TW'
    ]
  },

  requires: {
    "db": undefined,
    "app-service": {
      // this is the default implementation used for provided services
      impl: '@sap/cds-runtime/lib/cds-services/services/Service.js'
    },
    "auth": {
      '[development]': { kind: 'mocked-auth' },
      '[production]': { kind: 'JWT-auth' }
    },
    "dummy-auth": {
      strategy: 'dummy',
    },
    "mocked-auth": {
      strategy: 'mock',
      users: {
        alice: { roles: ['admin'] },
        bob: { roles: ['builder'] },
        '*': true
      }
    },
    "JWT-auth": {
      strategy: 'JWT',
    },
    destinations: undefined,
    monitoring: undefined,
    logging: undefined,
    audit: undefined,
    "sql": {
      '[development]': { kind: 'sqlite', credentials: { database: ':memory:' } },
      '[production]': { kind: 'hana' },
    },
    "sqlite": {
      // use: '@sap/cds-sqlite', // enforce using sqlite when accessed through <sql>
      use: 'sqlite', // enforce using sqlite when accessed through <sql>
      credentials: { database: 'sqlite.db' },
      impl: '@sap/cds-runtime/lib/sqlite/Service.js'
    },
    "hana": {
      // use: '@sap/cds-hana', // enforce using hana when accessed through <sql>
      use: 'hana', // enforce using hana when accessed through <sql>
      impl: '@sap/cds-runtime/lib/hana/Service.js'
    },
    "rest": {
      impl: '@sap/cds-runtime/lib/rest/service.js'
    },
    "odata": {
      impl: '@sap/cds-runtime/lib/rest/service.js'
    },
    "local-messaging": {
      impl: '@sap/cds-runtime/lib/messaging/service.js'
    },
    "file-based-messaging": {
      outbox: true,
      impl: "@sap/cds-runtime/lib/messaging/file-based.js",
      credentials: { file:'~/.cds-msg-box' }
    },
    "enterprise-messaging": {
      outbox: true,
      impl: '@sap/cds-runtime/lib/messaging/enterprise-messaging.js'
    },
    'message-queuing': {
      outbox: true,
      impl: '@sap/cds-runtime/lib/messaging/message-queuing.js'
    },
    "composite-messaging": {
      impl: "@sap/cds-runtime/lib/messaging/composite.js"
    }
  },

  odata: {
    flavors: {
      v2: {
        version: 'v2',
        // containment: false,
        // struct:false,
          // refs:false,
            // proxies:false,
  },
      v4: {
        version: 'v4',
        // containment: false,
        // struct:false,
          // refs:false,
            // proxies:false,
      },
      w4: { // for ODM with Fiori clients
        version: 'v4',
        containment:true,
        structs:true,
          refs:false, //> proxies:false,
        safe_ax:true,
      },
      x4: { // for A2X APIs
        version: 'v4',
        containment:true,
        structs:true,
          refs:true, //> proxies:true,
        safe_ax:true,
      },
    },
    version: 'v4',
    // following is to support code completion only
    structs: undefined,
    refs: undefined,
    proxies: undefined,
    containment: undefined,
    safe_ax: undefined,
  },

  sql: {
    names: 'plain' // or 'quoted', or 'hdbcds'
  },

  hana: {
    'deploy-format': 'hdbcds' // or 'hdbtable'
  },

  build: {
    target: 'gen'
  },

  mtx: {
    api: {
      model: true,
      provisioning: true,
      metadata: true
    },
    domain: '__default__'
  },

  cdsc: {
    // toSql: { associations: 'joins' },
    // newCsn: true,
  },

  query: {
    limit: {
      max: 1000
    }
  },

}
