// @sap/cds-runtime/lib/db/sql-builder/BaseBuilder.js
const _slugify = name => name.replace(/::/g, '__').replace(/\./g, '_')
const cds = global.cds || require('@sap/cds/lib')

const quotingStyles = {
  quoted: (name, delimiter) => `${delimiter}${name}${delimiter}`,
  plain: name => _slugify(name),
  'all-upper': (name, delimiter) => `${delimiter}${_slugify(name.toUpperCase())}${delimiter}`,
  'all-lower': (name, delimiter) => `${delimiter}${_slugify(name.toLowerCase())}${delimiter}`,
  bracketed: name => `[${name}]`
}

/**
 * BaseBuilder class should not be instantiated. *
 */
class BaseBuilder {
  /**
   * The base class constructor for the builders.
   * If the options parameter is not specified, " are used as delimiter and ? as placeholders.
   *
   * @param {object} obj - The CQN object used for the insert operation
   * @param {object} [options] - The configuration object.
   * @param {string} [options.delimiter] - The delimiter string.
   * @param {string} [options.placeholder] - The placeholder for prepared statement.
   * @param {object} csn - The csn object
   */
  constructor (obj, options, csn) {
    this._obj = obj
    this._csn = csn
    this._options = { ...this.getDefaultOptions(), ...options }
    this._quotingStyle = cds.env.sql.names || 'plain'
    this._validateQuotingStyle()
  }

  getDefaultOptions () {
    return {
      placeholder: '?',
      delimiter: '"',
      user: 'ANONYMOUS',
      now: { sql: 'NOW ()' }
    }
  }

  _validateQuotingStyle () {
    let type = typeof this._quotingStyle
    if (type !== 'string' || !quotingStyles.hasOwnProperty(this._quotingStyle)) {
      type = type !== 'string' ? `Type ${type}` : `"${this._quotingStyle}"`
      throw new Error(`Quoting style: ${type} is not supported. Allowed strings: "quoted", "plain".`)
    }
  }

  _quoteElement (element) {
    return quotingStyles[this._quotingStyle](element, this._options.delimiter, this._csn)
  }
}

module.exports = BaseBuilder
