// @sap/cds-runtime/lib/messaging/message-queuing-utils/management.js
const authorizedRequest = require('../common-utils/authorizedRequest')
const cds = global.cds || require('@sap/cds/lib')
const LOG = cds.log('messaging')

const _oa2 = oauth2 => ({
  clientid: oauth2.clientId,
  clientsecret: oauth2.clientSecret,
  tokenendpoint: oauth2.tokenUrl
})
const addSubscription = (queueName, topicPattern, msgSrv) =>
  authorizedRequest({
    method: 'PUT',
    uri: msgSrv.optionsClient.management.url,
    path: `/v1/management/queues/${encodeURIComponent(queueName)}/subscriptions/topics/${encodeURIComponent(
      topicPattern
    )}`,
    oa2: _oa2(msgSrv.optionsClient.management.auth.oauth2),
    attemptInfo: () => LOG._debug && LOG.debug('Add subscription', { topic: topicPattern, queue: queueName }),
    rejectString: `Subscription "${topicPattern}" could not be added to queue "${queueName}".`,
    msgSrv
  })

const putQueue = (queueName, msgSrv) =>
  authorizedRequest({
    method: 'PUT',
    uri: msgSrv.optionsClient.management.url,
    path: `/v1/management/queues/${encodeURIComponent(queueName)}`,
    oa2: _oa2(msgSrv.optionsClient.management.auth.oauth2),
    dataObj: msgSrv.queueConfig,
    attemptInfo: () => LOG._debug && LOG.debug('Put queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be created.`,
    msgSrv
  })

const deleteQueue = (queueName, msgSrv) =>
  authorizedRequest({
    method: 'DELETE',
    uri: msgSrv.optionsClient.management.url,
    path: `/v1/management/queues/${encodeURIComponent(queueName)}`,
    oa2: _oa2(msgSrv.optionsClient.management.auth.oauth2),
    attemptInfo: () => LOG._debug && LOG.debug('Delete queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be deleted.`,
    msgSrv
  })

module.exports = {
  addSubscription,
  putQueue,
  deleteQueue
}
