// @sap/cds-runtime/lib/hana/pool.js
const cds = global.cds || require('@sap/cds/lib')

const pool = require('@sap/cds-foss')('generic-pool')
const hana = require('./driver')

const _require = require('../common/utils/require')

let im

function multiTenantInstanceManager (db = cds.env.requires.db) {
  const creds =
    (db.credentials && (db.credentials.get_managed_instance_url || db.credentials.sm_url) && db.credentials) ||
    _require('@sap/xsenv').serviceCredentials(db.vcap || { label: 'managed-hana' })

  if (!creds || typeof creds !== 'object' || !(creds.get_managed_instance_url || creds.sm_url)) {
    throw Object.assign(new Error('No or malformed Managed HANA credentials'), { credentials: creds })
  }

  // new instance manager
  return new Promise((resolve, reject) => {
    // REVISIT: better cache settings? current copied from old cds-hana...
    // note: may need to be low for mtx tests -> configurable?
    const opts = Object.assign(creds, {
      cache_max_items: 1,
      cache_item_expire_seconds: 1
    })
    _require('@sap/instance-manager').create(opts, (err, res) => {
      if (err) return reject(err)
      resolve(res)
    })
  })
}

function singleTenantInstanceManager (db = cds.env.requires.db) {
  const creds = db.credentials || _require('@sap/xsenv').serviceCredentials(db.vcap || { label: 'hana' })

  if (!creds || typeof creds !== 'object' || !creds.host) {
    throw Object.assign(new Error('No or malformed HANA credentials'), { credentials: creds })
  }

  // mock instance manager
  return {
    get: (_, cb) => {
      cb(null, { credentials: creds })
    }
  }
}

async function credentials4 (tenant, credentials) {
  if (!im) {
    const opts = credentials ? { credentials } : undefined
    im = cds.env.requires.db.multiTenant ? await multiTenantInstanceManager(opts) : singleTenantInstanceManager(opts)
  }

  return new Promise((resolve, reject) => {
    im.get(tenant, (err, res) => {
      if (err) return reject(err)
      if (!res) return reject(new Error(`There is no instance for tenant "${tenant}"`))
      resolve(res.credentials)
    })
  })
}

function factory4 (creds, tenant) {
  return {
    create: function () {
      return hana.__connect(creds, tenant)
    },
    destroy: function (client) {
      return hana.__disconnect(client)
    },
    validate: function (client) {
      return hana.__isConnected(client)
    }
  }
}

/*
 * default generic-pool config
 */
const config = { min: 0, max: 100, testOnBorrow: true }

// REVISIT: copied from old cds-hana
const _getMassagedCreds = function (creds) {
  if (!('ca' in creds) && creds.certificate) {
    creds.ca = creds.certificate
  }
  if ('encrypt' in creds && !('useTLS' in creds)) {
    creds.useTLS = creds.encrypt
  }
  if ('hostname_in_certificate' in creds && !('sslHostNameInCertificate' in creds)) {
    creds.sslHostNameInCertificate = creds.hostname_in_certificate
  }
  if ('validate_certificate' in creds && !('sslValidateCertificate' in creds)) {
    creds.sslValidateCertificate = creds.validate_certificate
  }
  return creds
}

const _getPoolConfig = function () {
  const { pool: poolConfig } = cds.env.requires.db

  const mergedConfig = Object.assign({}, config, poolConfig)

  // defaults
  if (!poolConfig) {
    if (process.env.NODE_ENV === 'production') {
      mergedConfig.acquireTimeoutMillis = 0.3 * 1000
    } else {
      mergedConfig.acquireTimeoutMillis = 10 * 1000
    }
    mergedConfig.softIdleTimeoutMillis = 30 * 1000
  }

  // if evictionRunIntervalMillis is not set specifically, set to 2x of idleTimeoutMillis or softIdleTimeoutMillis
  if (!('evictionRunIntervalMillis' in mergedConfig)) {
    mergedConfig.evictionRunIntervalMillis =
      2 * (mergedConfig.idleTimeoutMillis || mergedConfig.softIdleTimeoutMillis || 30 * 1000)
  }

  // if numTestsPerEvictionRun is not set specifically, set to ~20% of min-max delta
  if (!('numTestsPerEvictionRun' in mergedConfig) && mergedConfig.max - mergedConfig.min > 0) {
    mergedConfig.numTestsPerEvictionRun = Math.ceil((mergedConfig.max - mergedConfig.min) / 5)
  }

  return mergedConfig
}

const pools = new Map()

async function pool4 (tenant, credentials) {
  if (!pools.get(tenant)) {
    pools.set(
      tenant,
      new Promise((resolve, reject) => {
        credentials4(tenant, credentials)
          .then(creds => {
            resolve(pool.createPool(factory4(_getMassagedCreds(creds), tenant), _getPoolConfig()))
          })
          .catch(e => {
            // delete pools entry if fetching credentials failed
            pools.delete(tenant)
            reject(e)
          })
      }).then(p => {
        pools.set(tenant, p)
        return p
      })
    )
  }
  if ('then' in pools.get(tenant)) {
    pools.set(tenant, await pools.get(tenant))
  }

  return pools.get(tenant)
}

module.exports = {
  acquire: (tenant, credentials) => {
    return pool4(tenant, credentials).then(p => p.acquire())
  },
  release: client => {
    return pool4(client._tenant).then(p => p.release(client))
  },
  drain: async tenant => {
    if (!pools.get(tenant)) {
      return
    }
    const p = await pool4(tenant)
    pools.delete(tenant)
    await p.drain()
    await p.clear()
  }
}
