// @sap/cds-runtime/lib/messaging/enterprise-messaging-utils/management.js
const authorizedRequest = require('../common-utils/authorizedRequest')
const cds = global.cds || require('@sap/cds/lib')
const LOG = cds.log('messaging')

const putQueue = (queueName, msgSrv) =>
  authorizedRequest({
    method: 'PUT',
    uri: msgSrv.optionsClient.management.uri,
    path: `/hub/rest/api/v1/management/messaging/queues/${encodeURIComponent(queueName)}`,
    oa2: msgSrv.optionsClient.management.oa2,
    dataObj: msgSrv.queueConfig,
    attemptInfo: () => LOG._debug && LOG.debug('Put queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be created.`,
    msgSrv
  })

const addSubscription = (queueName, topicPattern, msgSrv) =>
  authorizedRequest({
    method: 'PUT',
    uri: msgSrv.optionsClient.management.uri,
    path: `/hub/rest/api/v1/management/messaging/queues/${encodeURIComponent(
      queueName
    )}/subscriptions/${encodeURIComponent(topicPattern)}`,
    oa2: msgSrv.optionsClient.management.oa2,
    attemptInfo: () => LOG._debug && LOG.debug('Add subscription', { topic: topicPattern, queue: queueName }),
    rejectString: `Subscription "${topicPattern}" could not be added to queue "${queueName}".`,
    msgSrv
  })

const deleteQueue = (queueName, msgSrv) =>
  authorizedRequest({
    method: 'DELETE',
    uri: msgSrv.optionsClient.management.uri,
    path: `/hub/rest/api/v1/management/messaging/queues/${encodeURIComponent(queueName)}`,
    oa2: msgSrv.optionsClient.management.oa2,
    attemptInfo: () => LOG._debug && LOG.debug('Delete queue', { queue: queueName }),
    rejectString: `Queue "${queueName}" could not be deleted.`,
    msgSrv
  })

module.exports = {
  addSubscription,
  putQueue,
  deleteQueue
}
